### ISSUE TITLE
Correct way to include 'secret' for IOS-based devices?

### ISSUE THREAD

**Issue by bdlamprecht** [nornir-automation/nornir#364]
Let me begin by saying I'm **loving** the idea of being able to move away from being dependent on the _ever-so-popular_ DSL called `ansible`, so having this python-based framework for automation of networking devices is great!

I've successfully created a `nornir` object with ~8,500 devices using the `netbox` inventory plugin which is working as I want (see PR #361 requesting help for creating updated tests for that plugin).

Also, I've defined a `transform_function` for adding credentials to those devices (in the `config.yaml` file which is passed to `InitNornir`) which is also working well, but only for Juniper devices.

When I try to connect to Cisco devices, I get the following traceback:
```
ValueError: Failed to enter enable mode. Please ensure you pass the 'secret' argument to ConnectHandler.
```

This makes sense, as there is no `secret` that I've defined in the `transform_function`.  I've tried to follow the [documentation](https://nornir.readthedocs.io/en/latest/plugins/connections/index.html#module-nornir.plugins.connections.napalm) for that argument, but all to no avail.

The following is what I've tried in my `transform_function` code:
```
def add_credentials(host):
    host.username = "my_username"
    host.password = "my_password"
    host.extras = {"secret":"my_password"}
```

but I get the following traceback using that example:
```
AttributeError: 'Host' object has no attribute 'extras'
```

I've found some previous issues which have been opened against v1 of `nornir` which no longer apply as the structure has changed significantly.

In short, what I'm I doing wrong and is there a "best practice" for accomplishing this?

**Comment by bdlamprecht** [comment]
Copying here from the `NetworkToCode` Slack channel:

I tried to do this as a workaround:
```
for device in ios_devices.inventory.hosts:
    nr.inventory.hosts[device].connection_options = {
        "napalm": {
            "extras": {
                "secret": nr.inventory.hosts[device].password
            }
        }
    }
```
and it _appears_ to have added the correct structure:
```
print(nr.inventory.hosts["my_device"].connection_options)
{'napalm': {'extras': {'secret': 'my_password'}}}
```
However, when I try and run the napalm getter for `facts`, I get this traceback:
```
vvvv napalm_get ** changed : False vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv ERROR
Traceback (most recent call last):
  File "/home/bl839s/nornir/nornir/core/task.py", line 67, in start
    r = self.task(self, **self.params)
  File "/home/bl839s/nornir/nornir/plugins/tasks/networking/napalm_get.py", line 49, in napalm_get
    device = task.host.get_connection("napalm", task.nornir.config)
  File "/home/bl839s/nornir/nornir/core/inventory.py", line 293, in get_connection
    conn = self.get_connection_parameters(connection)
  File "/home/bl839s/nornir/nornir/core/inventory.py", line 230, in get_connection_parameters
    hostname=r.hostname if r.hostname is not None else self.hostname,
AttributeError: 'dict' object has no attribute 'hostname'
```

**Comment by dbarrosop** [comment]
You can't just pass a dictionary there, however, something like this should work:

```
from nornir.core.inventory import ConnectionOptions

def add_credentials(host):
    host.username = "my_username"
    host.password = "my_password"
    host.connection_options["napalm"] = ConnectionOptions(extras={"secret": "my_secret"})
```

I think we need to write a howto for this as you are not the first person to ask.

Alternatively, if you want to set it for all of your hosts you could also do it by adding it to defaults:

```
nr.inventory.defaults.connection_options = ConnectionOptions(extras={"secret": "my_secret"})
```

**Comment by bdlamprecht** [comment]
Thanks for responding. I'm glad to hear that I'm not the only one struggling to get this to work.
I didn't think I was unique in that manner but after the exchanges in Slack yesterday, I was starting to believe otherwise.

Even though you said it was "from the top of [your] head", I tried the example you provided above.
Unfortunately, the following traceback was generated upon attempting to run it:
```
vvvv napalm_get ** changed : False vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv ERROR
Traceback (most recent call last):
  File "/home/bl839s/nornir/nornir/core/task.py", line 67, in start
    r = self.task(self, **self.params)
  File "/home/bl839s/nornir/nornir/plugins/tasks/networking/napalm_get.py", line 49, in napalm_get
    device = task.host.get_connection("napalm", task.nornir.config)
  File "/home/bl839s/nornir/nornir/core/inventory.py", line 302, in get_connection
    extras=conn.extras,
  File "/home/bl839s/nornir/nornir/core/inventory.py", line 352, in open_connection
    configuration=configuration,
  File "/home/bl839s/nornir/nornir/plugins/connections/napalm.py", line 50, in open
    connection = network_driver(**parameters)
TypeError: __init__() got an unexpected keyword argument 'secret'

^^^^ END napalm_get ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The good part is that the `nr` object was actually created this time instead of failing due to the _previously_ **bad** format of `transform_function` which I _guessed_ at.

Now, when I try and look at the `connection_options` for each host using this code:
```
for device in ios_devices.inventory.hosts:
    print(f"{device} ({nr.inventory.hosts[device].platform})")
    print(f"\t{nr.inventory.hosts[device].hostname}")
    print(f"\t{nr.inventory.hosts[device].username}")
    print(f"\t{nr.inventory.hosts[device].password}")
    print(f"\t{nr.inventory.hosts[device].connection_options}")
```
I get the following output:
```
switch_1 (ios)
	9.88.240.21
	my_username
	my_password
	{'napalm': <nornir.core.inventory.ConnectionOptions object at 0x7f6717656150>}
switch_2 (ios)
	9.88.240.29
	my_username
	my_password
	{'napalm': <nornir.core.inventory.ConnectionOptions object at 0x7f67176561a8>}
```
Regarding this comment from above...
>I think we need to write a howto for this 

I would volunteer to do it if someone would show me how to get it to work properly.

**Comment by dbarrosop** [comment]
> TypeError: __init__() got an unexpected keyword argument 'secret'

Right, sorry. What gets assigned to "extras" is whatever the driver is expecting. So in your case is:

```
    host.connection_options["napalm"] = ConnectionOptions(extras={"optional_args": {...})
```

**Comment by bdlamprecht** [comment]
That worked perfectly, thank you!

**Comment by dbarrosop** [comment]
Closing this one as everything seems to be done :)
